ddht/tools/lru_sql_dict.py

```diff
@@ -0,0 +1,271 @@
+import pickle
+import sqlite3
+from typing import Any, Iterator, MutableMapping, NamedTuple, Tuple, Union
+
+from eth_utils.toolz import first
+
+CREATE_CACHE_QUERY = """
+    CREATE TABLE IF NOT EXISTS cache (
+        id INTEGER NOT NULL PRIMARY KEY,
+        key BLOB NOT NULL UNIQUE,
+        value BLOB NOT NULL,
+        pref INTEGER UNIQUE,
+        nref INTEGER UNIQUE
+    )
+"""
+
+CREATE_MAP_QUERY = """
+    CREATE TABLE IF NOT EXISTS map (
+        key BLOB NOT NULL PRIMARY KEY,
+        pointer INTEGER NOT NULL UNIQUE
+    )
+"""
+
+DEFAULT_CACHE_SIZE = 100000
+
+
+class Node(NamedTuple):
+    id: int
+    key: Union[str, bytes]
+    value: Any
+    pref: int
+    nref: int
+
+
+class LRUSQLDict(MutableMapping[Union[str, bytes], Any]):
+    """
+    SQLite3-backed dictionary that implements an LRU cache.
+    """
+
+    def __init__(
+        self, conn: sqlite3.Connection, cache_size: int = DEFAULT_CACHE_SIZE
```

Comment:
Two thoughts.

1. I don't think this should be have default size
2. Allowing `None` as the default size would allow for using this without bounds... but we don't necessarily have an immediate use case for that.

Reply:
I've updated the default size to be `None`. My initial intent with the default size was to avoid a db-backed dict from being created with some cache-size, and then later on that same db being instantiated with a smaller cache size - messing up the cache integrity. Is this a case to be worried about / handled?